tools/expr/expr.1
.\"	$Id: expr.1,v 1.2 2000/08/14 16:26:08 amai Exp $
.\"	$NetBSD: expr.1,v 1.9 1995/04/28 23:27:13 jtc Exp $
.\"
.\" Written by J.T. Conklin <jtc@netbsd.org>.
.\" Public domain.
.\"
.Dd July 3, 1993
.Dt EXPR 1
.Os
.Sh NAME
.Nm expr
.Nd evaluate expression
.Sh SYNOPSIS
.Nm expr
.Ar expression
.Sh DESCRIPTION
The
.Nm
utility evaluates
.Ar expression
and writes the result on standard output.
All operators are separate arguments to the
.Nm
utility.
Characters special to the command interpreter must be escaped.
.Pp
Operators are listed below in order of increasing precedence.
Operators with equal precedence are grouped within { } symbols.
.Bl -tag -width indent
.It Ar expr1 Li | Ar expr2
Returns the evaluation of
.Ar expr1
if it is neither an empty string nor zero;
otherwise, returns the evaluation of
.Ar expr2 .
.It Ar expr1 Li & Ar expr2
Returns the evaluation of
.Ar expr1
if neither expression evaluates to an empty string or zero;
otherwise, returns zero.
.It Ar expr1 Li "{=, >, >=, <, <=, !=}" Ar expr2
Returns the results of integer comparison if both arguments are integers;
otherwise, returns the results of string comparison using the locale-specific
collation sequence.
The result of each comparison is 1 if the specified relation is true,
or 0 if the relation is false.
.It Ar expr1 Li "{+, -}" Ar expr2
Returns the results of addition or subtraction of integer-valued arguments.
.It Ar expr1 Li "{*, /, %}" Ar expr2
Returns the results of multiplication, integer division, or remainder of integer-valued arguments.
.It Ar expr1 Li : Ar expr2
The
.Ql \&:
operator matches
.Ar expr1
against
.Ar expr2 ,
which must be a regular expression.  The regular expression is anchored
to the beginning of the string with an implicit
.Ql ^ .
.Pp
If the match succeeds and the pattern contains at least one regular
expression subexpression
.Dq "\e(...\e)" ,
the string corresponding to
.Dq "\e1"
is returned;
otherwise, the matching operator returns the number of characters matched.
If the match fails and the pattern contains a regular expression subexpression
the null string is returned;
otherwise, returns 0.
.El
.Pp
Parentheses are used for grouping in the usual manner.
.Sh EXAMPLES
Add 1 to the variable
.Va a :
.Pp
.Dl a=`expr $a + 1`
.Pp
Return the filename portion of a pathname stored
in variable
.Va a .
The
.Ql //
characters act to eliminate ambiguity with the division operator.
.Pp
.Dl expr "//$a" Li : '.*/\e(.*\e)'
.Pp
Return the number of characters in variable
.Va a :
.Pp
.Dl expr $a Li : '.*'
.Pp
.Sh DIAGNOSTICS
The
.Nm
utility exits with one of the following values:
.Pp
.Bl -tag -width Ds -compact
.It 0
The expression is neither an empty string nor 0.
.It 1
The expression is an empty string or 0.
.It 2
The expression is invalid.
.It >2
An error occurred (such as memory allocation failure).
.El
.Sh SEE ALSO
.Xr test 1
.Sh STANDARDS
The
.Nm
utility conforms to
.St -p1003.2 .


ChangeLog
		[$Id: ChangeLog,v 1.34 2000/11/27 18:13:42 arnd Exp $]
 
2000/03/14 LIBREVISION 0.01
	Initial version (arnd)
2000/03/22
	added fixcrcr utility and missing Makefiles in libext/uni (veit)

2000/03/24 LIBREVISION 0.02
	big change: (arnd)
	beautified fixcrcr.c, more docs, added syslog/drand/valloc/dlfcn,
	more headers, added some functions from EMX stat.c, added some
	i387 libm asm funcs, added a math test vector
2000/03/29 
	additions: (arnd)
	signal,resource,time header files and man pages
	Updated CodeRules
	Added Id lines to various files
	Support for itimer and daemon
	Misc fixes

2000/04/03 LIBREVISION 0.03
	additions: (arnd)
	more BSD man2 docs, signal(3) docs
	some missing headers (limits, signal, cdefs, itimer, resource)
	signalx.c was missing in CVS
	popen.c
	fixed a few Makefiles
	more CVSIds in files

2000/04/17 LIBREVISION 0.04 (0.04a)
	additions: (arnd)
	more docs, several sys headers from BSD
	regex support, some more stdlib: a64l, quad, tsearch
	sysconf support
	test for popen
	renamed fixcrcr.exe to fixtxt.exe
2000/04/18 (0.04b)
	additions: (arnd)
	err* stuff, __volatile__ macro fixed
	several minor CVS-Id changes
	msginfo structure renamed (clash with macro)
	removed timext.h
	_getPosixOS2Root cast fixed

2000/04/26 LIBREVISION 0.05
	additions: (arnd)
	perror, vis, man pages
	declare several arrays as pointers rather than []
	various conformance tests for FP and regex
	finally remove timext.h :-)
	set several constants from param.h which where hardcoded before
	init some local variables in regcomp to NULL
	empty getloadavg
	added some assembler math functions (i387math.S) from libgcc
	regexp debug function

2000/05/02 LIBREVISION 0.06
	additions: (arnd, Stefan Neis, Ralf Barkow)
	i387math.S fixes for acos/asin/exp from FreeBSD (neis)
	move fcntl.h after sys/types.h (barkow) and sys/stat.h (arnd)
	man pages: poll,bitstring,cat*,get*,killpg,math,hz,intro9
	i386 specific float.h, ieee.h->ieeefp.h
	nls support,killpg declaration, 
	IPC and SHM fixed,zlib integration
	mmap from BSD removed, nls messages for different languages
	sysconf fixes
	tests for db,shm
2000/05/05
	Math update (Cephes Math Lib), addition of SYS-facitilities.txt
	moved drand
	don't use gcc builtins
	handling of infinity
	test for floatingpoint funcs, consistency test
2000/05/08
	Math update s_tanh.c (neis)
2000/05/12
	Fix to stdio and unistd headers to make C++ happy (neis)

2000/05/15 LIBREVISION 0.07
	additions: ctype, libgen,
	docs for mmap, man5 pages
	tools: the ultimate editor (porting example), pmake
2000/05/17
	Minor update (arnd)
	additions: 4.3 compatible regex, realpath, nice
	more manpages
	HowTo pages for some OS/2 stuff
2000/05/31
	Update rev007b (arnd)
	fixcrcr.c: added check to detect binaries (\0 bytes in text)
	changed librevision format in dll.c
	added several EDM/2 articles and moved html files to docs/html
	some more manpages
	More stuff to FAQ
	renamed some fdlibm functions to __fdlibm
	__STRICT_ANSI__ removed from some defines
	register in prototypes removed
	itimer management improved
	added getsubopt,l64a,mergesort,bm,strlcpy,strmode,strsep
	non-Posix string extensions: strcasecmp etc.
2000/06/13
	math.h _LIB_VERSION var moved (neis)
2000/07/06
	minor changes of defs in math.h and limits.h for LONG LONG support
	(arnd)

Alpha #1 released

2000/07/24
	missing stuff for tools/pmake (arnd)
2000/07/25
	missing newline character added in some headers (amai)
	OpenBSD version of db added (arnd)
	misc fixes: headers, sources, docs (arnd)
2000/07/28
	<poll.h> added, minor updates for docs (amai)

2000/07/29 LIBREVISION 0.09
	lockf() added: not functional yet (see fcntl() in EMX) (arnd)
	timezone stuff added (arnd)
        Minor overall fixed (Makefiles, docs) (amai)

Alpha #2 released

2000/08/06
	Update rev009b
	time.h added, mmap fixes (arnd)
	removed #warnings from some headers - will make autoconf happy (arnd)
2000/08/08
	Update rev009c
	signgam added to libm (arnd)
	plain text version of FAQ removed in favour of HTML version (arnd)
	Makefile fixes (arnd, amai)
	[minor fixes to Makefiles and documentation are not always
	 logged here!]

2000/08/14 LIBREVISION 0.0.10
	misc fixes for libext (move some stuff in according subdirectories,
	dirent stuff, ...)   (arnd)
	Add big tree of utilities to tools/
2000/08/14 LIBREVISION 0.0.11
	few minor fixes for libExt (arnd)
	Modifications for Makefiles, make use of .mk files (arnd)
	"Missing" stuff in tools/ from rev 010 (arnd)
	working copy of rename(); tests for rename() and stat() (amai)
2000/08/15 LIBREVISION 0.0.11a
	few minor fixes for libExt; cleanup of directories (arnd)
	add data files to src/share (arnd)
	misc fixes for tools/ tree; added find(1) (arnd)
2000/08/18 LIBREVISION 0.0.11b
	Misc changes and additions; more files moved location (arnd)
	(In an early state this release was messed up in CVS!)
2000/08/19 LIBREVISION 0.0.12
	Misc stuff as usual. Mainly Makfiles this time,
	introduce two dlls instead of the single one
      change directory structures a bit (arnd)
2000/08/19 LIBREVISION 0.0.12a
2000/08/21 LIBREVISION 0.0.12b
2000/08/23 LIBREVISION 0.0.12c
	Dirent headers, glob* + misc stuff (arnd)
2000/08/26 LIBREVISION 0.0.13
	Build up basic directory structures (arnd)
      Add calendar and rmdir to tools/ as well as some manpages (arnd)
      Fixes/additions to libext
2000/09/09 LIBREVISION 0.0.14
      Add asa and printenv to tools/ as well as some manpages and 
      docs (arnd).
      Fixes/additions to libExt (arnd):
	1) New directory libext/libc/sys for modules that emulate BSD system 
	calls in a safer and more compatible way. Moved several ones.
	2) sys/mm.c enhanced and prepared for later split that will be 
	necessary for linking into a run-time DLL. New function 
	mprotect(). Now mmap() works with MAP_SHARED.
	3) sys/statx.c enhanced: New function statfs(). Several headers 
	adapted: sys/vfs.h, sys/statfs.h, sys/mount.h.
2000/10/11 LIBREVISION 0.0.14a (arnd)
	more changes of mm.c (MAP_SHARED/MAP_FIXED better supported)

tools/m4/expr.c
/*	$OpenBSD: expr.c,v 1.8 1999/09/16 20:19:34 espie Exp $	*/
/*	$NetBSD: expr.c,v 1.7 1995/09/28 05:37:31 tls Exp $	*/

/*
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Ozan Yigit at York University.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
#if 0
static char sccsid[] = "@(#)expr.c	8.2 (Berkeley) 4/29/95";
#else
static char rcsid[] = "$OpenBSD: expr.c,v 1.8 1999/09/16 20:19:34 espie Exp $";
#endif
#endif /* not lint */

#include <sys/cdefs.h>
#include <err.h>
#include <stdio.h>
#include <ctype.h>
#include "mdef.h"
#include "extern.h"

/*
 *      expression evaluator: performs a standard recursive
 *      descent parse to evaluate any expression permissible
 *      within the following grammar:
 *
 *      expr    :       query EOS
 *      query   :       lor
 *              |       lor "?" query ":" query
 *      lor     :       land { "||" land }
 *      land    :       not { "&&" not }
 *	not	:	eqrel
 *		|	'!' not
 *      eqrel   :       shift { eqrelop shift }
 *      shift   :       primary { shop primary }
 *      primary :       term { addop term }
 *      term    :       exp { mulop exp }
 *	exp	:	unary { expop unary }
 *      unary   :       factor
 *              |       unop unary
 *      factor  :       constant
 *              |       "(" query ")"
 *      constant:       num
 *              |       "'" CHAR "'"
 *      num     :       DIGIT
 *              |       DIGIT num
 *      shop    :       "<<"
 *              |       ">>"
 *      eqrel   :       "="
 *              |       "=="
 *              |       "!="
 *      	|       "<"
 *              |       ">"
 *              |       "<="
 *              |       ">="
 *
 *
 *      This expression evaluator is lifted from a public-domain
 *      C Pre-Processor included with the DECUS C Compiler distribution.
 *      It is hacked somewhat to be suitable for m4.
 *
 *      Originally by:  Mike Lutz
 *                      Bob Harper
 */

#define EQL     0
#define NEQ     1
#define LSS     2
#define LEQ     3
#define GTR     4
#define GEQ     5
#define OCTAL   8
#define DECIMAL 10
#define HEX	16

static const char *nxtch;		       /* Parser scan pointer */

static int query __P((void));
static int lor __P((void));
static int land __P((void));
static int not __P((void));
static int eqrel __P((void));
static int shift __P((void));
static int primary __P((void));
static int term __P((void));
static int exp __P((void));
static int unary __P((void));
static int factor __P((void));
static int constant __P((void));
static int num __P((void));
static int geteqrel __P((void));
static int skipws __P((void));
static void experr __P((char *));

/*
 * For longjmp
 */
#include <setjmp.h>
static jmp_buf expjump;

/*
 * macros:
 *      ungetch - Put back the last character examined.
 *      getch   - return the next character from expr string.
 */
#define ungetch()       nxtch--
#define getch()         *nxtch++

int
expr(expbuf)
	const char *expbuf;
{
	int rval;

	nxtch = expbuf;
	if (setjmp(expjump) != 0)
		return FALSE;

	rval = query();
	if (skipws() == EOS)
		return rval;

	printf("m4: ill-formed expression.\n");
	return FALSE;
}

/*
 * query : lor | lor '?' query ':' query
 */
static int
query()
{
	int bool, true_val, false_val;

	bool = lor();
	if (skipws() != '?') {
		ungetch();
		return bool;
	}

	true_val = query();
	if (skipws() != ':')
		experr("bad query");

	false_val = query();
	return bool ? true_val : false_val;
}

/*
 * lor : land { '||' land }
 */
static int
lor()
{
	int c, vl, vr;

	vl = land();
	while ((c = skipws()) == '|') {
		if (getch() != '|')
			ungetch();
		vr = land();
		vl = vl || vr;
	}

	ungetch();
	return vl;
}

/*
 * land : not { '&&' not }
 */
static int
land()
{
	int c, vl, vr;

	vl = not();
	while ((c = skipws()) == '&') {
		if (getch() != '&')
			ungetch();
		vr = not();
		vl = vl && vr;
	}

	ungetch();
	return vl;
}

/*
 * not : eqrel | '!' not
 */
static int
not()
{
	int val, c;

	if ((c = skipws()) == '!' && getch() != '=') {
		ungetch();
		val = not();
		return !val;
	}

	if (c == '!')
		ungetch();
	ungetch();
	return eqrel();
}

/*
 * eqrel : shift { eqrelop shift }
 */
static int
eqrel()
{
	int vl, vr, eqrel;

	vl = shift();
	while ((eqrel = geteqrel()) != -1) {
		vr = shift();

		switch (eqrel) {

		case EQL:
			vl = (vl == vr);
			break;
		case NEQ:
			vl = (vl != vr);
			break;

		case LEQ:
			vl = (vl <= vr);
			break;
		case LSS:
			vl = (vl < vr);
			break;
		case GTR:
			vl = (vl > vr);
			break;
		case GEQ:
			vl = (vl >= vr);
			break;
		}
	}
	return vl;
}

/*
 * shift : primary { shop primary }
 */
static int
shift()
{
	int vl, vr, c;

	vl = primary();
	while (((c = skipws()) == '<' || c == '>') && getch() == c) {
		vr = primary();

		if (c == '<')
			vl <<= vr;
		else
			vl >>= vr;
	}

	if (c == '<' || c == '>')
		ungetch();
	ungetch();
	return vl;
}

/*
 * primary : term { addop term }
 */
static int
primary()
{
	int c, vl, vr;

	vl = term();
	while ((c = skipws()) == '+' || c == '-') {
		vr = term();

		if (c == '+')
			vl += vr;
		else
			vl -= vr;
	}

	ungetch();
	return vl;
}

/*
 * <term> := <exp> { <mulop> <exp> }
 */
static int
term()
{
	int c, vl, vr;

	vl = exp();
	while ((c = skipws()) == '*' || c == '/' || c == '%') {
		vr = exp();

		switch (c) {
		case '*':
			vl *= vr;
			break;
		case '/':
			if (vr == 0)
				errx(1, "division by zero in eval.");
			else
				vl /= vr;
			break;
		case '%':
			if (vr == 0)
				errx(1, "modulo zero in eval.");
			else
				vl %= vr;
			break;
		}
	}
	ungetch();
	return vl;
}

/*
 * <term> := <unary> { <expop> <unary> }
 */
static int
exp()
{
	int c, vl, vr, n;

	vl = unary();
	switch (c = skipws()) {

	case '*':
		if (getch() != '*') {
			ungetch();
			break;
		}

	case '^':
		vr = exp();
		n = 1;
		while (vr-- > 0)
			n *= vl;
		return n;
	}

	ungetch();
	return vl;
}

/*
 * unary : factor | unop unary
 */
static int
unary()
{
	int val, c;

	if ((c = skipws()) == '+' || c == '-' || c == '~') {
		val = unary();

		switch (c) {
		case '+':
			return val;
		case '-':
			return -val;
		case '~':
			return ~val;
		}
	}

	ungetch();
	return factor();
}

/*
 * factor : constant | '(' query ')'
 */
static int
factor()
{
	int val;

	if (skipws() == '(') {
		val = query();
		if (skipws() != ')')
			experr("bad factor");
		return val;
	}

	ungetch();
	return constant();
}

/*
 * constant: num | 'char'
 * Note: constant() handles multi-byte constants
 */
static int
constant()
{
	int i;
	int value;
	int c;
	int v[sizeof(int)];

	if (skipws() != '\'') {
		ungetch();
		return num();
	}
	for (i = 0; i < sizeof(int); i++) {
		if ((c = getch()) == '\'') {
			ungetch();
			break;
		}
		if (c == '\\') {
			switch (c = getch()) {
			case '0':
			case '1':
			case '2':
			case '3':
			case '4':
			case '5':
			case '6':
			case '7':
				ungetch();
				c = num();
				break;
			case 'n':
				c = 012;
				break;
			case 'r':
				c = 015;
				break;
			case 't':
				c = 011;
				break;
			case 'b':
				c = 010;
				break;
			case 'f':
				c = 014;
				break;
			}
		}
		v[i] = c;
	}
	if (i == 0 || getch() != '\'')
		experr("illegal character constant");
	for (value = 0; --i >= 0;) {
		value <<= 8;
		value += v[i];
	}
	return value;
}

/*
 * num : digit | num digit
 */
static int
num()
{
	int rval, c, base;
	int ndig;

	rval = 0;
	ndig = 0;
	c = skipws();
	if (c == '0') {
		c = skipws();
		if (c == 'x' || c == 'X') {
			base = HEX;
			c = skipws();
		} else {
			base = OCTAL;
			ndig++;
		}
	} else
		base = DECIMAL;
	for(;;) {
		switch(c) {
			case '8': case '9':
				if (base == OCTAL) 
					goto bad_digit;
				/*FALLTHRU*/
			case '0': case '1': case '2': case '3': 
			case '4': case '5': case '6': case '7':
				rval *= base;
				rval += c - '0';
				break;
			case 'A': case 'B': case 'C': case 'D': case 'E': case 'F':
				c = tolower(c);
			case 'a': case 'b': case 'c': case 'd': case 'e': case 'f':
				if (base == HEX) {
					rval *= base;
					rval += c - 'a' + 10;
					break;
				}
				/*FALLTHRU*/
			default:
				goto bad_digit;
		}
		c = getch();
		ndig++;
	}
bad_digit:
	ungetch();
	
	if (ndig == 0)
		experr("bad constant");
	
	return rval;
}

/*
 * eqrel : '=' | '==' | '!=' | '<' | '>' | '<=' | '>='
 */
static int
geteqrel()
{
	int c1, c2;

	c1 = skipws();
	c2 = getch();

	switch (c1) {

	case '=':
		if (c2 != '=')
			ungetch();
		return EQL;

	case '!':
		if (c2 == '=')
			return NEQ;
		ungetch();
		ungetch();
		return -1;

	case '<':
		if (c2 == '=')
			return LEQ;
		ungetch();
		return LSS;

	case '>':
		if (c2 == '=')
			return GEQ;
		ungetch();
		return GTR;

	default:
		ungetch();
		ungetch();
		return -1;
	}
}

/*
 * Skip over any white space and return terminating char.
 */
static int
skipws()
{
	int c;

	while ((c = getch()) <= ' ' && c > EOS)
		;
	return c;
}

/*
 * resets environment to eval(), prints an error 
 * and forces eval to return FALSE.
 */
static void
experr(msg)
	char *msg;
{
	printf("m4: %s in expr.\n", msg);
	longjmp(expjump, -1);
}
